Fixed: Invert overlay scale and rotate on flip

BeebBenjamin opened the pull request:
This fix sets the correct position for the overlay upon `viewport.flipped`, it also applies a `ScaleX(-1)` and inverts the degrees for correctly rotating the overlay with the canvas.

The overlay is flipped entirely, this may not be desirable from a content perspective, however, I think people using OSD should invert content themselves because I do not think it is the job of OSD.  This fix is at least a step in the right direction.

Fixes: #2450

iangilman reviewed (changes requested):
This is a big improvement!

Looks like we're not quite there yet, though. I grabbed this code and tested it with `test/demo/overlay.html` (in this repository); three of the overlays are behaving perfectly, but one is not. Can you take a look at that?

Also, as you said, it would probably be better if the overlay contents weren't flipped; presumably people will be using text in their overlays and won't want the text to become  unreadable because the image has been flipped. My concern with taking this as an interim step is that if we release it like this, people will add their own code to flip the contents back, and if we then later release a fix, it'll break that code. 

Let's at least do some thinking about what might be possible for fixing the content flipping. I see a few possibilities:

* Nest the overlay inside another div so we can scale the outer div and then reverse scale the inner one. 
* Do the flipping with some other technique than the scale (e.g. rotate the opposite direction and update the position as needed).
* Something else?

BeebBenjamin commented:
Do you mean the green box `Scaled Overlay` still has the wrong positioning?

BeebBenjamin commented:
The green example overlay seems to lose it's x positioning after rotation, before that it's fine...

iangilman commented:
Yes, the green box, and yes, it is fine until you rotate.

BeebBenjamin commented:
OK, I have fixed the Scaled overlay X position not being right on rotate.

BeebBenjamin commented:
As for the content, in some instances people (me) use images as overlays and so they want the contents flipped.  I like the idea of an inner div, perhaps with a configurable in the OSD config to flip this inner div?  So if you want it to match the outer div `matchOverlayScales = true` or something.

BeebBenjamin commented:
The last commit adds code to create an inner div from the `textContent` which is then inverted in the X scale compared to it's parent.

iangilman reviewed (changes requested):
Great! All 4 of the overlays are registered solidly now.

The nesting divs solution seems promising. I don't think it's quite there yet, though. Copying the `textContent` works for pure text, but it'll break any styles applied to the text. 

Perhaps it would be better to wrap the extra div around the given element, so the element itself is unchanged?

BeebBenjamin commented:
I've added some changes to wrap the innerHTML of the overlay element in a `<div>` upon initialisation. For my purposes I actually want to flip the content with the overlay, so I have added an OSD config value to turn off the content flipping.

iangilman commented on src/openseadragon.js:
```diff
@@ -230,6 +230,9 @@
   * @property {Boolean} [flipped=false]
   *     Initial flip state.
   *
+  * @property {Boolean} [overlayContentFlipped=false]
```
We need a good name for this, which I think is going to be a challenge. It needs to communicate "if the viewport is flipped, flip the contents of overlays back so they still read correctly" but in fewer words. Maybe something like `overlayFlipCompensation`.

Then the description below should be much more detailed to explain the purpose of this flag.

BeebBenjamin replied:
How about `overlayFlipReversal` or `overalScaleReversal`?

iangilman commented on src/openseadragon.js:
```diff
@@ -1337,6 +1340,7 @@ function OpenSeadragon( options ){
 
             // INITIAL FLIP STATE
             flipped:                    false,
+            overlayContentFlipped:      false,
```
We should think about what the correct default should be. I know in your case having the flip compensation off is correct, but what is the average person going to want? Or more specifically, what is the expected default behavior?

I guess I'm thinking having the compensation on by default makes sense... "I want my overlay text readable" is a pretty uncontroversial statement. If someone needs them to be flipped, that's more of a specific interest.

What do you think?

BeebBenjamin replied:
Yeah readable overlays by default.

iangilman commented on src/overlay.js:
```diff
@@ -254,19 +255,23 @@
                 this.onDraw(position, size, this.element);
             } else {
                 var style = this.style;
+                var outerElement = element.firstChild;
```
Isn't this the inner element, not the outer element?

BeebBenjamin replied:
Yes, it is.  Sorry, was tired!

iangilman reviewed (changes requested):
This is looking great! Thank you for going the extra mile to make this functionality work, even though it's outside of your use case. I really appreciate it.

Just a couple more details...

iangilman commented on src/openseadragon.js:
```diff
@@ -230,8 +230,10 @@
   *
-  * @property {Boolean} [overlayContentFlipped=false]
-  *     Initial overlay content flip state.
+  * @property {Boolean} [overlayFlipReversal=true]
```
I think we're getting there, but it still seems confusing. Is it going to reverse flipping the overlay entirely? Maybe it should be `overlayFlipContentReversal` or something. I know that's long, but it's kind of a specialized property so that's okay. I suppose `overlayFlipReadable` might be good... Gets to the point of the feature.

pearcetm replied:
Other options could be something like `overlayPreserveLTR`, 'overlayPreserveContentDirection' (`direction` being the name of the css property for this), or the even more verbose `overlayPreserve[...]WhenFlipped`. "Preserve" could also be swapped out for "Maintain" in any of the above.

BeebBenjamin replied:
I think `overlayPreserveContentDirection` is probably the best one so far.

iangilman reviewed (changes requested):
Still musing about the name. Everything else looks good except that you haven't renamed `overlayContentFlipped` in every instance.

iangilman reviewed (approved):
That's a great name! Thank you both for the brainstorming :)

And thank you @BeebBenjamin for this robust patch!

sinamcr7 commented:
hi, thanks for great work, I'm using annotorious for annotations and wanted to apply same for it, is it possible or should I ask it in annotorious repository?

iangilman commented:
@sinamcr7 That probably needs to happen in the Annotorious repository. Feel free to point at this patch if it is useful!
